library/manageiq_custom_attributes.py

```diff
@@ -0,0 +1,259 @@
+                   id=entity_id)
+            result = self.client.get(url)
+            return result.get('custom_attributes')
+        except Exception as e:
+            self.module.fail_json(msg="Failed to get {entity_type} custom attributes. Error: {error}".format(
+                entity_type=entity_type, error=e))
+
+    def add_custom_attributes(self, entity_type, entity_id, custom_attributes):
+        """ Returns the added custom attributes """
+        try:
+            url = '{api_url}/{entity_type}/{id}/custom_attributes'.format(
+                    api_url=self.api_url,
+                    entity_type=ManageIQCustomAttributes.supported_entities[entity_type],
+                    id=entity_id)
+            result = self.client.post(url, action='add', resources=custom_attributes)
+            self.changed = True
+            return result['results']
+        except Exception as e:
+            self.module.fail_json(msg="Failed to add the custom attributes. Error: {}".format(e))
+
+    def update_custom_attribute(self, entity_type, entity_id, ca, ca_href):
+        """ Returns the updated custom attributes """
+        try:
+            url = '{api_url}/{entity_type}/{id}/custom_attributes'.format(
+                    api_url=self.api_url,
+                    entity_type=ManageIQCustomAttributes.supported_entities[entity_type],
+                    id=entity_id)
+            ca_object = {'name': ca['name'], 'href': ca_href, 'value': ca['value']}
+            result = self.client.post(url, action='edit', resources=[ca_object])
+            self.changed = True
+            return result['results']
+        except Exception as e:
+            self.module.fail_json(msg="Failed to update the custom attribute {ca_name}. Error: {error}".format(ca_name=ca['name'], error=e))
+
+    def add_or_update_custom_attributes(self, entity_type, entity_name, custom_attributes):
+        """ Adds custom attributes to an entity in manageiq or updates the
+        attributes in case already exists
+
+        Returns:
+            the added or updated custom attributes, whether or not a change
+            took place and a short message describing the operation executed
+        """
+        added, updated = [], []
+        message = ""
+        # check if entity with the type and name passed exists in manageiq
+        entity_id = self.find_entity_by_name(entity_type, entity_name)
+        if not entity_id:  # entity doesn't exist
+            self.module.fail_json(
+                msg="Failed to add the custom attributes. {entity_type} {entity_name} does not exist".format(entity_type=entity_type, entity_name=entity_name))
+
+        entity_cas = self.get_entity_cas(entity_type, entity_id)
+        if not entity_cas:  # no custom attributes for this entity
+            added.extend(self.add_custom_attributes(entity_type, entity_id, custom_attributes))
+        else:
```

Comment:
I think you can drop this clause if you do `self.get_entity_cas(...) or []` or better yet make did `get_entity_cas()` itself return [].
Then you'd only call `add_custom_attributes()` from single place (which just happens for each of them if none existed).